src/solvers/dg_curved/containers_3d.jl

```diff
@@ -68,16 +68,76 @@ function calc_jacobian_matrix!(jacobian_matrix::AbstractArray{<:Any,6}, element,
 end
 
 
-# Calculate contravarant vectors, multiplied by the Jacobian determinant J of the transformation mapping.
-# Those are called Ja^i in Kopriva's blue book.
-function calc_contravariant_vectors!(contravariant_vectors::AbstractArray{<:Any,6}, element, jacobian_matrix)
-  # TODO: This needs to be adapted for actually curved meshes.
-  # For rectangular meshes, the contravariant_vectors are just the scaled unit vectors
-  fill!(view(contravariant_vectors, .., element), 0)
+# Calculate contravariant vectors, multiplied by the Jacobian determinant J of the transformation mapping,
+# using the invariant curl form.
+# These are called Ja^i in Kopriva's blue book.
+function calc_contravariant_vectors!(contravariant_vectors::AbstractArray{<:Any,6}, element, 
+                                     jacobian_matrix, node_coordinates, basis::LobattoLegendreBasis)
+  # The general form is
+  # Jaⁱₙ = 0.5 * ( ∇ × (Xₘ ∇ Xₗ - Xₗ ∇ Xₘ) )ᵢ  where (n, m, l) cyclic and ∇ = (∂/∂ξ, ∂/∂η, ∂/∂ζ)ᵀ
+
+  # Calculate the first summand of the cross product in each dimension
+  for dims in 0:2, j in 1:nnodes(basis), i in 1:nnodes(basis)
```

Comment:
Maybe you could implement an `eachnode(basis)` and replace the `1:nnodes(basis)` occurrences here and in other parts of the implementaiton?